### DISCUSSION TITLE
Text styling for output (ANSI code)

### DISCUSSION THREAD

**Discussion by Taitava** [Taitava/obsidian-shellcommands#328] [Ideas - done & released]
Sometimes shell command output might contain control characters that define colors and/or font style (e.g. bold, italic) for parts of the outputted text. This is called [ANSI code](https://en.wikipedia.org/wiki/ANSI_escape_code).

So far, the _Shell command_ plugin has displayed the any possibly outputted _ANSI code_ texts as-is. I.e. the plugin has had no knowledge of it. It has resulted in the following hard to read output messages:

Example shell command: `git -c color.ui=always status` (shows [Git](https://git-scm.com) version control status, with color-coding enabled)

Original, ugly output:
![kuva](https://user-images.githubusercontent.com/13002142/226112389-39810b76-40a0-4c7b-9a54-f55eb79c96e7.png)

Output after converting ANSI code to HTML:
![kuva](https://user-images.githubusercontent.com/13002142/226112571-83e2a2a6-86e5-4e52-8774-d67dfcf8f32d.png)

**Comment by Taitava** [comment]
This is now released, but still lacks documentation.

**Comment by Taitava** [comment]
Hi @OMSmolina, I'm using this discussion to reply to [your comment in issue 329](https://github.com/Taitava/obsidian-shellcommands/issues/329#issuecomment-1883601398)
> In output: `Ask after execution` Ansi colors not working correctly.
> 
> ![image](https://private-user-images.githubusercontent.com/6513086/295306623-3e1c5efe-1265-4c1b-b4af-c698b5786af1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDQ5ODc5OTEsIm5iZiI6MTcwNDk4NzY5MSwicGF0aCI6Ii82NTEzMDg2LzI5NTMwNjYyMy0zZTFjNWVmZS0xMjY1LTRjMWItYjRhZi1jNjk4YjU3ODZhZjEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDExMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAxMTFUMTU0MTMxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NGY0M2ZkZjBjOTIxNjFlNDIwNTdmYmU4MWRhN2U3YTJiOTYwMGQyMzdmZjcxNzVkMTk4ZDRmNDk0YzdjNTA4NiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.wIvJ4HjCF8b4O7W_ciyA-0utDTWJD9jhmunoYwHtBbs)

Yeah, the [Ask after execution](https://publish.obsidian.md/shellcommands/Output+handling/Output+channel+-+Ask+after+execution) modal is actually lagging behind as I didn't notice to pay enough attention to it when developing the ANSI code support. That said, at least the output is still correct. If you click for example the _Current file: bottom_ button in the modal, the outputted content with the HTML formatting will show up correctly in a currently open note file.

A way to solve this would be to make the _Ask after execution_ modal to have a read-only mode, which would show HTML formatting correctly.

**Reply by OMSmolina** [reply]
Hi @Taitava 

Yes, the current file or notification bubble prints the correct colors, but it will also be nice to have them in Ask after execution modal.

The read-only mode option sounds good, to me it should be the default option like any console output.

**Reply by Taitava** [reply]
> Yes, the current file or notification bubble prints the correct colors, but it will also be nice to have them in Ask after execution modal.

I agree.

> The read-only mode option sounds good, to me it should be the default option like any console output.

Yes, there will be a setting to decide which will be the default mode for _Ask after execution_: read-only or editable.

**Reply by OMSmolina** [reply]
Any news about this?

**Reply by Taitava** [reply]
Thanks for asking again 🙂 . Sorry, no news yet. This basically requires some redesigning of the _Ask after execution_ modal. I wrote some ideas now to #408 . You can subscribe to the discussion to get notified when there's some progress, but I can't promise when I'll be able to start it.

**Comment by Taitava** [comment]
# Ability to remove ANSI code

@edshamis [commented](https://github.com/Taitava/obsidian-shellcommands/issues/329#issuecomment-1913106009) in #329:
> It would be nice to have an option to just remove ANSI code

A great idea! I'll plan this for ~~SC `0.23.0` - so not for the next release (`0.22.0`), but after that. 👍~~

**Reply by Taitava** [reply]
Hi @edshamis ! Sorry for the delay, I only just started to briefly look into this ability to remove ANSI code.
- It looks like the `ansi_up` library the SC plugin uses does not support removing ANSI code completely.
- There has been a [feature request for this](https://github.com/drudru/ansi_up/issues/69), but it didn't get implemented at least back then.
- Another library was mentioned that could possible do it: [strip-ansi](https://github.com/chalk/strip-ansi).

I'll need to consider what to do here. Introducing yet another library would not be the most optimal solution, but I guess it's the best in this situation. I'm also doubtful about modifying `ansi_up` myself for this, if the modification would have no possibilities to become part of the official library.

So, unfortunately I need to remove this feature from the `0.23.0` milestone and postpone this. If the new library looks decent, I might lean in that direction in the future.
